### ISSUE TITLE
Proposal to change printchplenv interface

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#7827]
PR #7757 changed the behavior of printchplenv to enable --llvm compiles to go faster. But, it left some open questions about printchplenv modes and more generally how to adjust that script so it's less confusing to maintain.

The below proposed interface for printchplenv came out of a discussion I had with @ben-albrecht that is related to PRs #7815 and #7757. In particular, the present problem is that `printchplenv --simple` is [documented as a way to produce chplconfig files](https://chapel-lang.org/docs/master/usingchapel/chplenv.html#generating-configuration-files)  but now outputs additional variables (for use by the Chapel compiler) that aren't appropriate for chplconfig files. A less severe, but still relevant problem, is that `printchplenv --debug` now outputs these same extra variables.

It seems that we need to add a mode to `printchplenv` in order to resolve these issues, but the script is already unreasonably tangled and confusing, and so it seemed time to think through how it might more reasonably be structured and what might be a better set of command line options for the script to support.

The proposal below preserves the following behaviors:
 * printchplenv with no arguments works exactly as it does today
 * printchplenv --simple and printchplenv --overrides can still produce chplconfig files

The proposal below makes the following changes in order to better structure the options:
 * enables separate selection of what variables to print / how to filter / output format
 * removes --sh and --csh modes (these are no longer needed now that we have chplconfig)
 * removes --debug in favor of --all
 * printchplenv --simple and printchplenv --overrides no longer outputs CHPL_LLVM by default (to keep the option sets & defaults reasonable. I expect this to be low impact because CHPL_LLVM is automatically enabled if it is built, but now users could use e.g. --simple --all)

Note that `chpl` and Makefiles can still get the output they are used to now
 * Makefiles would use `printchplenv --all --internal --no-tidy --make`
 * chpl compiler would use `printchplenv --all --internal --no-tidy --simple`

### Proposed printchplenv -h output

```
Usage: printchplenv [options]

Print the current Chapel configuration. Arguments allow selectiion of
what gets printed [content], how it gets printed [format], and what gets
filtered [filter].

The default contents are: --runtime --launcher

Options:
  -h, --help   show this help message and exit

  [shortcut]
  --all        shortcut for --compiler --runtime --launcher

  [content]
  --compiler    Select variables describing the configuration of the compiler
  --runtime     Select variables describing the configuration of the runtime
  --launcher    Select variables describing the configuration of the launcher
  --internal    Select additional variables used during builds

  [filter]
  --[no-]tidy  (default) [don't] omit sub-variables irrelevant to the current
                configuration
  --anonymize  avoid printing machine specific details such as host name and
               file paths
  --overrides  only show variables that have been user supplied
               via environment or chplconfig

  [format]
  --pretty     (default) print variables in format: CHPL_KEY: VALUE
               indicating which options are set by environment variables (*)
               and which are set by configuration files (+)
  --simple     print variables in format: CHPL_KEY=VALUE
               output is compatible with chplconfig format
  --make       print variables in format: CHPL_MAKE_KEY=VALUE
  --path       print variables in format: VALUE1/VALUE2/...
```

####  --compiler    Select variables describing the configuration of the compiler
    CHPL_HOST_PLATFORM
    CHPL_HOST_COMPILER
    CHPL_LLVM
    CHPL_MAKE

####  --runtime     Select variables describing the configuration of the runtime
    CHPL_TARGET_PLATFORM
    CHPL_TARGET_COMPILER
    CHPL_TARGET_ARCH
    CHPL_LOCALE_MODEL
    CHPL_COMM
    CHPL_COMM_SUBSTRATE
    CHPL_GASNET_SEGMENT
    CHPL_TASKS
    CHPL_TIMERS
    CHPL_UNWIND
    CHPL_MEM
    CHPL_JEMALLOC
    CHPL_MAKE
    CHPL_ATOMICS
    CHPL_NETWORK_ATOMICS
    CHPL_GMP
    CHPL_HWLOC
    CHPL_REGEXP
    CHPL_AUX_FILESYS

####  --launcher    Select variables describing the configuration of the launcher

```
CHPL_HOST_PLATFORM
CHPL_HOST_COMPILER
CHPL_LOCALE_MODEL
CHPL_COMM
CHPL_TASKS
CHPL_LAUNCHER
CHPL_TIMERS
CHPL_UNWIND
CHPL_MEM
CHPL_ATOMICS
```

####  --internal    Select additional variables used during builds
    CHPL_JEMALLOC
    CHPL_HOST_MEM
    CHPL_ORIG_TARGET_COMPILER
    CHPL_RUNTIME_SUBDIR
    CHPL_LAUNCHER_SUBDIR
    CHPL_COMPILER_SUBDIR
    CHPL_GMP_UNIQ_CFG_PATH
    CHPL_HWLOC_UNIQ_CFG_PATH
    CHPL_JEMALLOC_UNIQ_CFG_PATH
    CHPL_LIBUNWIND_UNIQ_CFG_PATH
    CHPL_MASSIVETHREADS_UNIQ_CFG_PATH
    CHPL_QTHREAD_UNIQ_CFG_PATH
    CHPL_RE2_UNIQ_CFG_PATH
    CHPL_THIRD_PARTY_LINK_ARGS

**Comment by bradcray** [comment]
At a glance, this looks generally attractive to me.  Isn't `CHPL_MAKE` used at compile time as well as "runtime"?

**Comment by mppf** [comment]
@bradcray yes, I've updated it. I think it's also likely that --launcher needs more variables that aren't described above but I don't have the exact set ready at hand.

**Comment by ben-albrecht** [comment]
One issue that has come up is that `CHPL_MAKE` is part of the current default output (`--runtime --launcher`), but should not be part of `--runtime --path` output.

This has made @mppf and I reconsider the value of including `CHPL_MAKE` in the default `printchplenv` output in the first place... Are there cases where this variable has been useful to users as a default print variable?

**Comment by bradcray** [comment]
I think it's been useful for us to help users debug issues which are due to 'make' vs. 'gmake' issues/confusions.

**Comment by ben-albrecht** [comment]
> I think it's been useful for us to help users debug issues which are due to 'make' vs. 'gmake' issues/confusions.

OK - I can work around it.

**Comment by ben-albrecht** [comment]
Upon working on this, we realized `--launcher` contains many unnecessary variables, which may have initially been an oversight:

```
CHPL_TARGET_PLATFORM
CHPL_TARGET_COMPILER
CHPL_LOCALE_MODE
CHPL_COMM
CHPL_TASKS
CHPL_LAUNCHER
CHPL_TIMERS
CHPL_UNWIND
CHPL_MEM
CHPL_ATOMICS
```

e.g. 

```bash
# printchplenv --launcher --path  (after these changes)
> printchplenv --launcher 
darwin/clang/loc-flat/comm-none/tasks-qthreads/launch-none/tmr-generic/unwind-none/mem-jemalloc/atomics-intrinsics
```

After chatting with @gbtitus , we deduced the launcher is only impacted by:

```
CHPL_TARGET_PLATFORM
CHPL_TARGET_COMPILER
CHPL_LOCALE_MODE
CHPL_COMM
CHPL_TASKS
CHPL_LAUNCHER
```

which would look like:

```bash
# printchplenv --launcher --path  (after these changes)
> printchplenv --launcher  
darwin/clang/loc-flat/comm-none/tasks-qthreads/launch-none
```

Note `CHPL_TASKS` does not impact the launcher configuration today, but it could in the future.

Would appreciate someone else checking our work before proceeding (@ronawho? @bradcray?)

**Comment by gbtitus** [comment]
A couple of notes/edits:
- The launcher shouldn't depend on the target compiler.  It's built by the host compiler.
- In conversation with @ben-albrecht I had said that as of the ugni dynamic heap registration work the launcher depended on the target memory setting and so I was surprised not to see it in this list. But on second thought and after reviewing the dyn reg changes, that launcher dependence on the target memory setting actually drives an execution-time decision in the launcher and doesn't affect its build, so there's no need for the target mem setting to be in the launcher unique config path.

**Comment by ronawho** [comment]
>I think it's been useful for us to help users debug issues which are due to 'make' vs. 'gmake' issues/confusions.

@bradcray while deciding if we could close https://github.com/chapel-lang/chapel/issues/6547 (and responding to your question about the chpl-env-gen.h test failure) I found myself thinking that it doesn't make sense to include `CHPL_MAKE` in the output of `printchplenv --runtime`, which would also mean it wouldn't be included in the default printchplenv output. 

That makes sense to me, because we also hide `CHPL_HOST_PLATFORM` and `CHPL_HOST_COMPILER` in the default `printchplenv` output, and `CHPL_MAKE` seems even less important to me than those values. It's also not used by the runtime, so far as I know the only place that cares about `CHPL_MAKE` is the compiler. Thoughts?

**Comment by bradcray** [comment]
Nothing other than what I said above:  It seems to me that there have been times when users have run into confusions in which they've had multiple versions of `make` on their system (where one has been non-gmake compliant) and that knowing what the value of `CHPL_MAKE` was in those cases was helpful.  But perhaps such cases are obvious enough that we can ask them to provide a more verbose output from `printchplenv` to verify rather than relying on it being in the default set of variables(?).

**Comment by ronawho** [comment]
I can't remember any cases like that (though I don't support users that much, so you would definetly know better) so I'm not sure if the value of `CHPL_MAKE` would be useful or not.

I would imagine the failure mode would be somewhat obvious though, so I'm definetly in favor of hiding `CHPL_MAKE` by default, and only printing it with `--compiler` or `--all`, and if a user does run into issue and we need more info we can always ask for the `--all` output, which would get us `CHPL_MAKE` as well as the `CHPL_HOST_*` values

**Comment by mppf** [comment]
I agree with @ronawho that it'd be much nicer for CHPL_MAKE to be only a --compiler variable. In that setting, we'd ask people for `printchplenv --all` when debugging make issues.

As far as user experience with confusion of make vs gmake, I think we should try to get our main Makefile to emit a reasonable error if it's used with BSD Make. I don't know how possible that is, though.

**Comment by mppf** [comment]
It has been bugging me that we didn't just give a nice error for accidental use of BSD Make. So I made a PR that checks. PR #8006. Assuming you are OK with that PR, I think we should be able to remove CHPL_MAKE from the default output.

**Comment by ben-albrecht** [comment]
As a follow-on to this, we are going to make the following changes (after #8067):

- `CHPL_MAKE` --> `--internal`
- `CHPL_NETWORK_ATOMICS` --> `--internal`
